plugins/com.google.cloud.tools.eclipse.util/src/com/google/cloud/tools/eclipse/util/ArtifactRetriever.java

```diff
@@ -0,0 +1,196 @@
+                  XPath xpath = XPathFactory.newInstance().newXPath();
+                  NodeList versionNodes = (NodeList) xpath.evaluate(
+                      "/metadata/versioning/versions/version",
+                      document,
+                      XPathConstants.NODESET);
+                  Builder<ArtifactVersion> versions = ImmutableSortedSet.naturalOrder();
+                  for (int i = 0; i < versionNodes.getLength(); i++) {
+                    String versionString = versionNodes.item(i).getTextContent();
+                    versions.add(new DefaultArtifactVersion(versionString));
+                  }
+                  return versions.build();
+                }
+              });
+
+  private static final ArtifactRetriever INSTANCE = new ArtifactRetriever();
+  public static ArtifactRetriever defaultInstance() {
+    return INSTANCE;
+  }
+
+  /**
+   * Returns the latest published artifact version, or null if there is no such version.
+   */
+  public ArtifactVersion getLatestArtifactVersion(String groupId, String artifactId) {
+    return getLatestIncrementalVersion(idToKey(groupId, artifactId), null);
+  }  
+  
+  /**
+   * Returns the latest published artifact version in the version range, or null if there is no such
+   * version.
+   */
+  public ArtifactVersion getLatestArtifactVersion(String groupId, String artifactId, VersionRange range) {
+    return getLatestIncrementalVersion(idToKey(groupId, artifactId), range);
+  }
+
+  /**
+   * Returns the latest version of the specified artifact in the version range, or null if there is
+   * no such version.
+   */
+  private ArtifactVersion getLatestIncrementalVersion(String coordinates, VersionRange range) {
+    try {
+      ArtifactVersion latest = latestVersion.get(coordinates);
+      if (range == null || range.containsVersion(latest)) {
+        return latest;
+      }
+    } catch (ExecutionException ex) {
+      logger.log(
+          Level.WARNING,
+          "Could not retrieve latest version for artifact " + coordinates,
+          ex.getCause());
+    }
+
+    try {
+      NavigableSet<ArtifactVersion> allVersions = availableVersions.get(coordinates);
+      ArtifactVersion latest = getLatestInRange(range, allVersions);
+      return latest;
+    } catch (ExecutionException ex) {
+      logger.log(
+          Level.WARNING,
+          "Could not retrieve available versions for artifact " + coordinates,
+          ex.getCause());
+      return null;
+    }
+  }
+
+  private static ArtifactVersion getLatestInRange(
+      VersionRange versionRange, NavigableSet<ArtifactVersion> allVersions) {
+    for (ArtifactVersion version : allVersions.descendingSet()) {
+      if (versionRange.containsVersion(version)) {
+        return version;
+      }
+    }
+    return null;
+  }
+
+  private Document getMetadataDocument(String coordinates) throws CoreException {
+    String[] x = keyToId(coordinates);
+    String groupId = x[0];
```

Comment:
Is it worth having an internal wrapper class with two fields that has real names for groupId/artifactId?

If you don't think so, then document the form that `coordinates` is permitted to be in wherever it is consumed (at least for the first time, i.e. with this style of key the cache should specify the requirements of its key)